Observe when an app has been upgraded as well as first launch.

krider2010 opened the pull request:
In addition to the first launch then it’s useful to know if an app has been upgraded to present some extra information to the user or to remind them about signing up etc.

This was something we wanted as well as knowing the first ever launch so I’ve added it in.

The app version is injectable. Not only for testing but if the user wants a custom way of reporting that information. A fully usable default is provided.

Tests have also been added for the new functionality.

krider2010 committed:
Observe when an app has been upgraded.

In addition to the first launch then it’s useful to know if an app has been upgraded to present some extra information to the user, or remind them about signing up etc.

This was something we wanted as well as first launch so I’ve added it in.

pixeldock commented:
Good idea! Thanks for your PR, I'll review it as soon as possible.

pixeldock commented on Example/RxAppState_ExampleTests/RxAppStateTests.swift:
```diff
@@ -118,6 +120,119 @@ class RxAppStateTests: XCTestCase {
         XCTAssertEqual(firstLaunchArray, [true])
     }
     
+    func testisFirstLaunchOfNewVersionNewInstall() {
```
This is a really small unimportant detail, but would you mind using a capital letter after the `test` in your test method names? Just to have consistency with the existing test method names. So in this case it would be `testIsFirstLaunchOfNewVersionNewInstall()`

krider2010 replied:
Absolutely - thanks for spotting the typo 👍

pixeldock commented on Pod/Classes/UIApplication+Rx.swift:
```diff
@@ -54,14 +54,36 @@ public func ==(lhs: AppState, rhs: AppState) -> Bool {
+    
+//    public static func getCurrentAppVersion(_ handler: ((Void) -> String?) = { _ in return Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String}) -> String? {
+//        return handler()
+//    }
```
Could you please remove theses commented lines?

krider2010 replied:
Doh - my bad!

pixeldock reviewed (commented):
Great PR! Please have a look at my comments and let me know what you think.

pixeldock commented on Pod/Classes/UIApplication+Rx.swift:
```diff
@@ -54,14 +54,36 @@ public func ==(lhs: AppState, rhs: AppState) -> Bool {
+     Allows for the app version to be stored by default in the main bundle from `CFBundleShortVersionString` or
+     a custom implementation per app.
+     */
+    public static var getCurrentAppVersion: ((Void) -> String?) = { _ in return Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String}
```
I would suggest renaming this to `currentAppVersion`. Having a property name starting with `get` sounds a bit strange IMHO.

Is there a particular reason that you use a closure for this property and not just the optional String like this?

`public static var currentAppVersion = Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String`

You could then inject the value just be setting the string to your desired value:

`RxAppState.currentAppVersion = "4.2"`

krider2010 replied:
Agreed on the renaming.

As for having it as a function, I was messing around with the version being calculated from a bunch of things. But thinking on it, that can still be done e.g. `RxAppState.currentAppVersion = someFunctionThatReturnsAString(possibleSomeParams)`

I'll make it a String.

krider2010 committed:
Fixes from PR review

- renamed test functions to be correct (my bad)
- removed dead code
- made the `currentAppVersion` an injectable `String?` rather than closure

krider2010 commented:
@pixeldock I've made all those changes. Thanks for highlighting them. Anything else, just let me know.

pixeldock reviewed (approved):
Thanks for the changes, I'll happily merge this into master. Thanks for the contribution!
